### DISCUSSION TITLE
Quick and dirty git integration (lazygit)

### DISCUSSION THREAD

**Discussion by smpallen99** [zed-industries/zed#8485] [Show and tell]
I just found tasks in zed last night. That make lazygit integration a breeze. Great feature zed.dev team! I'm probably addressing the folks that are using the vim bindings.

[LazyGit](https://github.com/jesseduffield/lazygit) as a terminal based Git client that uses vim bindings for navigation and control. 

Once that's installed, here is the zed task that will open it. You might need to install a zed version from the preview channel to get the [Tasks](https://github.com/zed-industries/zed/blob/main/docs/src/tasks.md) feature.

Open tasks with zed::OpenTasks action ("zed: open tasks" in the command palette) and add the following:

```  
[
   {
      "label": "Lazygit",
      "command": "lazygit",
      "args": [],
      "env": {},
      "use_new_terminal": true,
      "allow_concurrent_runs": false
  }
]
```

I have it mapped to `<leader>gg` as is fairly standard in neovim. 

```
[
  {
     "context": "Editor && vim_mode == normal && vim_operator == none && !VimWaiting", 
     "bindings": {
         // ...
        "space g g": ["workspace::SendKeystrokes", "alt-shift-t enter"]
   }
]
```

This solution isn't perfect. Here are a couple issues:

* I can't figure out how to send a command after the terminal opens with the binding above. I would like to have the  terminal open with LazyGit and zoom it. There is a limit with executing an async task and sending a command afterward. The workaround is to manually type `shift-escape` after `space g g`.
* When you quit LazyGit, the empty terminal tab must be manually closed. The workaround is to use `cmd-w` to close the active  pane.

**Comment by jamesgecko** [comment]
Is there a way to open a specific task from a key binding? It seems like there must be, as [the docs mention `$ZED_ROW` and others](https://zed.dev/docs/tasks). But I can't get it to work with `"space l": ["task::Spawn", "Lazygit"],`, it always opens the task runner.

**Reply by osiewicz** [reply]
Hey, we don't support binds like that right now, though it sounds relatively straightforward; I will work on it tomorrow. :)

**Reply by jamesgecko** [reply]
Wow, thank you!

**Reply by osiewicz** [reply]
Here we go: https://github.com/zed-industries/zed/pull/9064

**Reply by baldwindavid** [reply]
This is great. The workaround I am using for that currently is using the send keystrokes command...

```json
      "space t": [
        "workspace::SendKeystrokes",
        ": task:spawn enter Test Under Cursor enter"
      ],
```

This will be much cleaner. Thanks!

**Reply by failable** [reply]
@osiewicz Can we have an option to not to show the terminal when running a task? I setup a task to run `magit` in Emacs and I don't run Emacs in terminal but in another GUI. So the popup terminal is useless for me in this situation and I have to close it every time.

```
"ctrl-x m": [
        "task::Spawn",
        {
          "task_name": "Magit"
        }
      ]
```


```
{
    "label": "Magit",
    "command": "osascript",
    "args": [
      "-e",
      "'activate application \"Emacs\"'",
      "-e",
      "'do shell script \"/usr/local/bin/emacsclient -n -u -e \\\"(magit-status)\\\"\"'"
    ],
    "use_new_terminal": true
  }
```

**Reply by ivanlauhk** [reply]
As Tasks works now, I can setup a keybind to launch a specific task.
There is a `use_new_terminal` option in Tasks, but is it possible to make it open the way like `workplace: new center terminal`?
I want it to open in the main window instead of having to drag it from the bottom.

**Comment by baldwindavid** [comment]
@smpallen99 I posted some similar experiences in https://github.com/zed-industries/zed/discussions/9067 - Great to see these options available now.

**Comment by thuvasooriya** [comment]
I can't get space g g keybinding to work in zed, what could I possibly do wrong.
```
  {
    "context": "Editor && vim_mode == normal && !VimWaiting && !menu",
    "bindings": {
      // put key-bindings here if you want them to work only in normal mode
      "space g g": ["task:spawn", "lazygit"]
    }
  }
```

**Reply by floriansenat** [reply]
Does this work?
```jsonc
 "bindings": {
      // put key-bindings here if you want them to work only in normal mode
      "space g g": ["task::Spawn", { "task_name": "Lazygit" }]
 }
```

**Reply by thuvasooriya** [reply]
yes indeed! thanks. i just recently checked the docs also, seems they've updated it.

i wish there was a easy config to open lazygit as in Toggle zoom and it ends the task automatically and closed when i press q.
current process looks like ,

1. spawning the task with space g g  
2. shift esc to zoom  
3. do your thing and press q
4. close the terminal window with cmd w, if you have multiple terminals running good luck or you press ctrl ` i guess

it's kind of messed up ig. i looked around to find if there is any way to 
1. open a task in a new window 
2. a task terminal to exit automatically 
or just 
3. chain zed actions in general to do these things 
but i haven't found any. let me know if there is something i missed. thanks again.

edit: anyway to do this with SendKeystrokes?

**Reply by baldwindavid** [reply]
Tasks are great for one-off tests and such, but just not quite well-suited to this sort of thing. I use SendKeystrokes for a center terminal...

```json
      "space g g": [
        "workspace::SendKeystrokes",
        ": new center terminal enter lazygit enter"
      ],
```

This is adapted from my attempts to leverage a netrw-ish file explorer in https://github.com/zed-industries/zed/discussions/9067

You can then use the tab switcher to access as needed if it's in the same pane.

**Reply by thuvasooriya** [reply]
so cool works awesome! thanks!

**Reply by schoblaska** [reply]
I tweaked this to `": new center terminal enter lazygit space && space exit enter"` so that the terminal pane closes itself after I exit lazygit.

**Comment by mdrahiem** [comment]
For anyone looking for step by step implementation,

https://github.com/zed-industries/zed/issues/4367#issuecomment-2141358892

by @baldwindavid

**Reply by thuvasooriya** [reply]
> For anyone looking for step by step implementation,
> 
> [#4367 (comment)](https://github.com/zed-industries/zed/issues/4367#issuecomment-2141358892)
> 
> by @baldwindavid

i recommend you check out other implementations in this chat, having lazy git in a tiny terminal window is not ideal (at least for me), there are better implementations above than this

**Reply by mdrahiem** [reply]
Sorry, I didn't understand which other implementations you are talking about. Could you please redirect me to specific message/implementation you are referring to? I see the whole thread has `lazyGit` implementation but nothing else.

Also I see https://github.com/zed-industries/zed/discussions/8485#discussioncomment-9340239 but this doesn't work for me. When I press `space g g` the text `:newcenterterminallazygit` is pasted at my cursor but nothing else. Am I missing something here?

**Reply by thuvasooriya** [reply]
hmm that's strange. are you using vim mode? 
cause the method you mentioned didn't work for me and yes i was mentioning this https://github.com/zed-industries/zed/discussions/8485#discussioncomment-9340239
it works as i expect it to work as i'm using vim mode
maybe there is a better way to do this but i've put it in a bunch of places to make it work for me
keymap.json
```json
  {
    "context": "EmptyPane || SharedScreen",
    "bindings": {
      "space g g": [
        "workspace::SendKeystrokes",
        ": new center terminal enter lazygit space enter"
      ],
    }
  },
  {
    "context": "ProjectPanel && not_editing",
    "bindings": {
      "space g g": [
        "workspace::SendKeystrokes",
        ": new center terminal enter lazygit space enter"
      ],
    }
  },
  {
    "context": "Editor && vim_mode == normal && !VimWaiting && !menu",
    "bindings": {
      // only in normal mode
      "space g g": [
        "workspace::SendKeystrokes",
        ": new center terminal enter lazygit space enter"
      ],
...
    }
  },
  ```
  
  tasks.json
  ```json
  [
  {
    "label": "lazygit",
    "command": "lazygit",
    "args": [],
    "env": {},
    //"cwd": "/path/to/working/directory",
    "use_new_terminal": true,
    "allow_concurrent_runs": false,
    "reveal": "always"
  }
]
  ```

**Reply by baldwindavid** [reply]
@mdrahiem That original message https://github.com/zed-industries/zed/discussions/8485#discussioncomment-9686580 in which I'm attributed points to a message that is not written by me. The workflow I use is mentioned right in this discussion at https://github.com/zed-industries/zed/discussions/8485#discussioncomment-9340239

```json
  {
    "context": "Editor && VimControl && !VimWaiting && !menu",
    "bindings": {
      "space g g": [
        "workspace::SendKeystrokes",
        ": new center terminal enter lazygit enter"
      ],

  },
```

fwiw, I also make it available for an empty screen...

```json
  {
    "context": "EmptyPane || SharedScreen",
    "bindings": {
      "space g g": [
        "workspace::SendKeystrokes",
        ": new center terminal enter lazygit enter"
      ]
    }
  },
```

Note that there is no "task" involved here at all. I am just sending the keystrokes to open a center terminal and additional keystrokes to type the word `lazygit` followed by enter. This is the same as opening via a terminal outside Zed.

> When I press space g g the text :newcenterterminallazygit is pasted at my cursor but nothing else. Am I missing something here?

From the sounds of it I suspect you have added as a global keybinding and kicked off perhaps from the terminal. 🤷

**Reply by mdrahiem** [reply]
> hmm that's strange. are you using vim mode?

@thuvasooriya You're right. I didn't enable vim mode. When I did, it worked well. Thanks.

@baldwindavid Sorry for the wrong attribution. I see it's a different username. I got confused. Thanks for the clear explanation. I understand that to make key bindings work, I need to enable Vim mode.

**Comment by Blankeos** [comment]
How to change the `open` command of lazygit to open inside Zed?

**Reply by theherk** [reply]
You can do

```yaml
os:
  edit: "zed {{filename}}"
  editAtLine: "zed {{filename}}:{{line}}"
```

in your `~/.config/lazygit/config.yml`. You can surely do the same with the open command.

**Comment by theherk** [comment]
This is my current configuration for this combining all that I've learned here and bumps along the way. This is a bit more precise because it specifies bash. I use fish by default and that has problems here. Unclear why, see #20588. Also, it closes on `q`.

edit: This is updated based on two improvements present at 38c0aa303e8c. First, users of fish no longer need to specify another shell. Even better, thanks to @mikayla-maki's awesome change in #22004, one can now run this in the center pane.

### tasks.json

```json
[
  {
    "label": "lazygit",
    "command": "lazygit",
    "working_directory": "${ZED_WORKTREE_ROOT}",
    "hide": "on_success",
    "use_new_terminal": true,
    "allow_concurrent_runs": false
  }
]
```

### keymap.json

```json
[
  {
    "context": "Editor && vim_mode == normal && !VimWaiting && !menu",
    "bindings": {
      "space g g": ["task::Spawn", { "task_name": "lazygit", "target": "center"}]
    }
  }
]
```

**Reply by smedegaard** [reply]
I use fish. What precisely is not working with fish here?

**Reply by theherk** [reply]
This is working, but not with the shell as fish; at least not for me. You can see my explanation in the linked bug report. If it works for you with fish, let me know. You can peruse my configuration if you have ideas why at theherk/commons.

**Comment by pidanou** [comment]
Anyone knows how Lazygit integration on Neovim (LazyVim distro) [https://github.com/folke/snacks.nvim](https://github.com/folke/snacks.nvim) can open the UI so fast? it seems like in this package that the terminal was already live (i can't see the fastfetch/neofetch text for example).

It seems like in the background there is always a terminal running, and it is just shown when needed

**Reply by danieljimenez-mycase** [reply]
Mine was only slow because of my zsh config. So I use bash for mine lazygit.
```
  {
    "label": "start lazygit",
    "command": "lazygit -p $ZED_WORKTREE_ROOT",
    "hide": "on_success",
    "shell": {
      "program": "sh"
    }
  },
```

**Comment by pidanou** [comment]
After updating to `0.165.4`, `"space g g": [
        "workspace::SendKeystrokes",
        ": new center terminal enter lazygit space && space exit enter"
      ],
` no longer works, it just opens the terminal and types nothing

**Comment by baldwindavid** [comment]
@pidanou fyi, this no longer works as of last week's preview release and looks like no plans for it to work again. See #21610 and #21584 

Appears opening in the bottom pane via task is about the closest now.

**Reply by pidanou** [reply]
That's a shame, this was the only way to get a decent dev experience for git tui.
This breaking might send me back Neovim...

**Reply by theherk** [reply]
The other method given [here for example](https://github.com/zed-industries/zed/discussions/8485?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDk2MzEzMDkyMTI6TURRNlZYTmxjak0yTURVd056Zz0%3D#discussioncomment-11255047) does still work. Slightly different interface, but works.

**Reply by pidanou** [reply]
Yes but there are 2 issues: 
- cannot open the full terminal, which makes lazygit less usable
- it doesn't close the panel on quit

I went back to Neovim for few hours and then I switched back to Zed just for the scrolling butter smoothness.

But yea there are still some very very nice features on Neovim that I will be missing (leap and telescope)

**Reply by baldwindavid** [reply]
The task route works, but after this stopped working I have started to just manually open a center terminal, run `lazygit`, and then just keep the tab around. That this feature stopped working has been much more problematic for trying to use a center terminal for a netrw-like experience like #9067

**Reply by theherk** [reply]
@pidanou You're so right about Neovim. I'm an avid user too, but am working with Zed primarily for smooth scrolling. Wezterm, Ghostty, and Contour all have tasks open to implement smooth scrolling but no progress that I know of since it will be a complex task in terminals. But I'm testing Zed for other reasons too, like CRDT.

Anyway, on topic, the panel does close on quit. You're right it doesn't open full though. I actually open it then have `⌘ ⏎` set to full screen. But then hitting `q` does close the panel.

**Reply by baldwindavid** [reply]
Yup, maximizing is easy enough with a keyboard shortcut; just a flow thing. I think all of these issues would go away if we could target a task at a center terminal, but not sure if that will ever be a thing.

**Reply by theherk** [reply]
Would you look at what just landed; can't wait to try it out. #22004

**Reply by baldwindavid** [reply]
Yes, this is so awesome! Works perfect.

**Comment by Stewart86** [comment]
just wanna throw it out there for people using Hyprland / WM where we use window class parameter to float the window, this method works even better than Zed built-in terminal as I think the border drawing characters for lazygit are drawn better with kitty or any modern terminal, also felt a lot more integrated rather than a hack this way too. 

the command just execute your terminal of choice (kitty for me) with it floating then call lazygit within that terminal. Once you are done, when closing, with `"hide":"on_success"` field, it will hide the spawning terminal within Zed.

```json
// keymap.json
{
    "space g g": ["task::Spawn", { "task_name": "lazygit", "target": "center" }]
}

// tasks.json
[
    {
        "label": "lazygit",
        "command": "kitty --class 'dotfiles.floating' -e lazygit",
        "args": [],
        "env": {},
        "use_new_terminal": true,
        "allow_concurrent_runs": false,
        "working_directory": "${ZED_WORKTREE_ROOT}",
        "hide": "on_success",
        "reveal": "always"
    }
]
```

**Comment by CX330Blake** [comment]
Oh gosh I love lazygit lol

**Comment by danieljimenez-mycase** [comment]
For reference here is what my configs look like for zed + lazygit. Inspired by LazyVim defaults. Currently there is a bug that if a long run task is still running. Pressing the keybinding will not focus on lazygit.

tasks.json (I use bash to make it faster)
```
[
  {
    "label": "start lazygit",
    "command": "lazygit -p $ZED_WORKTREE_ROOT",
    "hide": "on_success",
    "shell": {
      "program": "sh"
    }
  },
  {
    "label": "start lazygit file log",
    "command": "lazygit",
    "args": ["-f", "\"$ZED_RELATIVE_FILE\""],
    "hide": "on_success",
    "shell": {
      "program": "sh"
    }
  },
]
```
keybindings.json
 ```
 {
    "context": "vim_mode == normal && !menu",
    "bindings": {
      "space g g": [
        "task::Spawn",
        { "task_name": "start lazygit", "reveal_target": "center" }
      ],
      "space g f": [
        "task::Spawn",
        { "task_name": "start lazygit file log", "reveal_target": "center" }
      ],
    }
  },
```


lazygit config. Let me edit and quit with escape.
```
quitOnTopLevelReturn: true
os:
  edit: "zed {{filename}}"
  editAtLine: "zed {{filename}}:{{line}}"

```

**Reply by Xeitor** [reply]
when using this config:

> ```
> quitOnTopLevelReturn: true
> os:
>   edit: "zed {{filename}}"
>   editAtLine: "zed {{filename}}:{{line}}"
> ```

after editing a file and returning to the lazygit task, lazygit asks for enter to return
```
╰─ lazygit

+ bash -c zed "...."

Press enter to return to lazygit
```

any ideas on how to prevent this?

**Reply by Xeitor** [reply]
found it:

adding this `promptToReturnFromSubprocess: false` to the root of `~/.config/lazygit/config.yml` does the trick

**Comment by aphexlog** [comment]
Thanks, fits perfectly
